aziot/src/init.rs

```diff
@@ -0,0 +1,948 @@
+	LocalCa,
+
+	#[display(fmt = "Issued dynamically from an EST server")]
+	Est,
+}
+
+#[derive(Clone, Copy, Debug, derive_more::Display)]
+enum EstIdSource {
+	#[display(fmt = "Stored on the filesystem (an \"EST ID\" certificate)")]
+	Preloaded,
+
+	#[display(fmt = "Issued dynamically from the EST server using a different client certificate (a \"bootstrap EST ID\" certificate)")]
+	Bootstrap,
+}
+
+#[derive(Clone, Copy, Debug, derive_more::Display)]
+enum EstTrustedCert {
+	#[display(fmt = "My EST server's server certificate can be validated with just the CA certificates already installed on my OS")]
+	Os,
+
+	#[display(fmt = "My EST server's server certificate can only be validated with a separate CA certificate.")]
+	Separate,
+}
+
+/// This trait exists to read lines from stdin without directly using `std::io::StdinLock`,
+/// because reading secrets (symmetric keys, etc) requires disabling echo on the tty directly which can't be done while stdin is locked.
+///
+/// Making a trait also allows it to be mocked for tests where the input comes from files.
+trait Reader {
+	/// Prints the given prompt string, reads a line from the user and appends the response to the given line buffer.
+	///
+	/// Unlike `std::io::BufRead::read_line`, this does not include a trailing newline. Specifically, empty input is returned as an empty string.
+	/// EOF is indicated by returning a `Err(std::io::ErrorKind::UnexpectedEof)` instead.
+	fn read_line(&mut self, prompt: &str, line: &mut String) -> std::io::Result<usize>;
+
+	/// Prints the given prompt string, reads a secret from the user and appends the response to the given line buffer.
+	///
+	/// Unlike `std::io::BufRead::read_line`, this does not include a trailing newline. Specifically, empty input is returned as an empty string.
+	/// EOF is indicated by returning a `Err(std::io::ErrorKind::UnexpectedEof)` instead.
+	fn read_secret(&mut self, prompt: &str, line: &mut String) -> std::io::Result<usize>;
+}
+
+struct Stdin {
+	prompter: rustyline::Editor<()>,
+}
+
+impl Reader for Stdin {
+	fn read_line(&mut self, prompt: &str, line: &mut String) -> std::io::Result<usize> {
+		loop {
+			match self.prompter.readline(prompt) {
+				Ok(response) => {
+					line.push_str(&response);
+					return Ok(response.len());
+				},
+
+				Err(rustyline::error::ReadlineError::Io(err)) => return Err(err),
+
+				Err(rustyline::error::ReadlineError::Eof) |
+				Err(rustyline::error::ReadlineError::Interrupted) => return Err(std::io::ErrorKind::UnexpectedEof.into()),
+
+				Err(rustyline::error::ReadlineError::Utf8Error) => println!("Input could not be parsed as UTF-8"),
+
+				Err(rustyline::error::ReadlineError::Errno(err)) => return Err(std::io::Error::new(std::io::ErrorKind::Other, err)),
+
+				Err(err) => return Err(std::io::Error::new(std::io::ErrorKind::Other, err)),
+			}
+		}
+	}
+
+	fn read_secret(&mut self, prompt: &str, line: &mut String) -> std::io::Result<usize> {
```

Comment:
theoretically, you could remove the rpassword dependency and re-use rustyline: https://github.com/kkawakam/rustyline/pull/266/files

I haven't checked how much extra "security" rpassword offers, but if it's "good enough" to simply replace the password chars with '*', we might be able to cut down on a dependency

Reply:
`rpassword` disables the echo attribute on the TTY in the standard way. `rustyline` instead puts the TTY into raw mode and uses the highlighter to emit the keystrokes. Both approaches are basically the same, so it should be fine.